pkg/cache/v3/delta.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package cache
+
+import (
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/envoyproxy/go-control-plane/pkg/log"
+	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
+)
+
+// Respond to a delta watch with the provided snapshot value. If the response is nil, there has been no state change.
+func respondDelta(request *DeltaRequest, value chan DeltaResponse, st *stream.StreamState, snapshot Snapshot, log log.Logger) *RawDeltaResponse {
+	resp, err := createDeltaResponse(request, st, snapshot)
+	if err != nil {
+		if log != nil {
+			log.Errorf("Error creating delta response: %v", err)
+		}
+		return nil
+	}
+
+	// Only send a response if there were changes
+	if len(resp.Resources) > 0 || len(resp.RemovedResources) > 0 {
+		if log != nil {
+			log.Debugf("node: %s, sending delta response with resources: %v removed resources %v wildcard: %t",
+				request.GetNode().GetId(), resp.Resources, resp.RemovedResources, st.IsWildcard)
+		}
+		value <- resp
+		return resp
+	}
+	return nil
+}
+
+func createDeltaResponse(request *DeltaRequest, st *stream.StreamState, snapshot Snapshot) (*RawDeltaResponse, error) {
+	resources := snapshot.GetResources((request.TypeUrl))
+
+	// variables to build our response with
+	nextVersionMap := make(map[string]string)
+	filtered := make([]types.Resource, 0, len(resources))
+	toRemove := make([]string, 0)
+
+	// Wildcard requests originate from CDS or LDS. If detect one, we want to track all resources as well as track them in the version map
```

Comment:
`If we are handling a wildcard request, we want to respond with all resources`?

I'm not sure what the "as well as track them in the version map" refers to

Reply:
Ah I guess that comment is redundant. I'll remove